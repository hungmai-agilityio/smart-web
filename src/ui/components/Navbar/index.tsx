import { NavItem } from '@/ui/components/Navbar/Item';

interface NavbarProps {
  items: Array<{ name: string; url?: string }>;
  styles?: string;
}

export const Navbar = ({ items, styles }: NavbarProps) => {
  return (
    <nav className={styles}>
      {items.map((item, index) => (
        <NavItem key={index} name={item.name} url={item.url} />
      ))}
    </nav>
  );
};
