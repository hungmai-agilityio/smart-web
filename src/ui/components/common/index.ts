export * from '@/ui/components/common/Button';
export * from '@/ui/components/common/Dropdown';
export * from '@/ui/components/common/Icon';
export * from '@/ui/components/common/Photo';
export * from '@/ui/components/common/Text';
export * from '@/ui/components/common/StatsRow';
