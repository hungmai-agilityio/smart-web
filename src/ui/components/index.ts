export * from '@/ui/components/Hero';
export * from '@/ui/components/NavItem';
export * from '@/ui/components/Portfolio';
export * from '@/ui/components/Pagination';
export * from '@/ui/components/Skeleton/Card';
export * from '@/ui/components/Skeleton/Image';
export * from '@/ui/components/common';
