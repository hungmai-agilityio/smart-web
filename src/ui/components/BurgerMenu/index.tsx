'use client';

import Image from 'next/image';
import { useRef, useState } from 'react';
import { Navbar } from '@/ui/components';
import { useOutsideClick } from '@/hooks';

interface BurgerProps {
  items: Array<{ name: string; url?: string }>;
}

export const BurgerMenu = ({ items }: BurgerProps) => {
  const menuRef = useRef<HTMLDivElement>(null);
  const [menuOpen, setMenuOpen] = useState<boolean>(false);

  const toggleMenu = () => {
    setMenuOpen(!menuOpen);
  };

  useOutsideClick({
    ref: menuRef,
    handler: () => setMenuOpen(false)
  });

  return (
    <div ref={menuRef}>
      <div
        className="md:hidden flex items-center cursor-pointer"
        onClick={toggleMenu}
      >
        <Image src="/menu.svg" alt="menu" width={16} height={14} priority />
      </div>

      <Navbar
        items={items}
        styles={`md:flex gap-7 ${menuOpen ? 'block' : 'hidden'}
          md:relative absolute top-full left-0 w-full md:w-auto z-10 transition-all ease-in-out
          bg-white p-5 drop-shadow-lg md:drop-shadow-none lg:mr-28 md:mr-8`}
      />
    </div>
  );
};
