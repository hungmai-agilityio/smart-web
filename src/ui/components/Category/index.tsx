'use client';

import { memo, useCallback } from 'react';
import { useRouter, useSearchParams } from 'next/navigation';

// Constants
import { abel } from '@/constants';

// Components
import { Dropdown, ItemProps } from '@/ui/components';

interface CategoryProps {
  categories: string[];
  queryParam: string;
}

export const Category = memo(({ categories, queryParam }: CategoryProps) => {
  const router = useRouter();
  const searchParams = useSearchParams();

  const currentCategory = searchParams.get(queryParam);

  // Update category and URL when a new category is selected
  const handleCategorySelect = useCallback(
    (category: string) => {
      const params = new URLSearchParams(searchParams.toString());
      params.set(queryParam, category);
      router.push(`?${params.toString()}`, { scroll: false });
    },
    [searchParams, router, queryParam]
  );

  const dropdownItems: ItemProps[] = categories.map((category) => ({
    label: category,
    value: category
  }));

  return (
    <>
      <ul
        className={`${abel.className} my-8 md:flex hidden lg:gap-48 gap-16 justify-center`}
      >
        {categories.map((category, index) => (
          <li
            className={`uppercase cursor-pointer hover:text-second ${currentCategory === category ? 'text-second' : ''}`}
            key={index}
            onClick={handleCategorySelect.bind(null, category)}
          >
            {category}
          </li>
        ))}
      </ul>
      <div className="md:hidden flex justify-center my-8">
        <Dropdown
          items={dropdownItems}
          selectedValue={currentCategory!}
          onSelect={(value) => handleCategorySelect(value)}
        />
      </div>
    </>
  );
});
