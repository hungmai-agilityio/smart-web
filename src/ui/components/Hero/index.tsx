import { ReactNode } from 'react';

// Constants
import { TYPE } from '@/constants';

// Components
import { Button, Photo } from '@/ui/components/common';

interface HeroProps {
  imageUrl: string;
  children?: ReactNode;
  buttonName?: string;
}

export const Hero = ({ imageUrl, children, buttonName }: HeroProps) => {
  return (
    <div className="relative w-full h-hero">
      <Photo src={imageUrl} alt="Hero Background" autoSize priority />
      <div className="absolute  inset-0 flex flex-col items-center justify-center">
        {children}
        {buttonName && (
          <div className="mt-10">
            <Button variant={TYPE.PRIMARY} name={buttonName} />
          </div>
        )}
      </div>
    </div>
  );
};
