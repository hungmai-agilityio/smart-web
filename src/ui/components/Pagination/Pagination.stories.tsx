import { Meta, StoryObj } from '@storybook/react';

// Components
import { Pagination } from '@/ui/components/Pagination';

const meta: Meta<typeof Pagination> = {
  component: Pagination,
  tags: ['autodocs'],
  parameters: {
    layout: 'centered',
    nextjs: {
      appDirectory: true
    }
  },
  argTypes: {
    pageCount: {
      description: 'Total number of pages'
    },
    currentPage: {
      description: 'The current active page number'
    }
  }
};

export default meta;

type Story = StoryObj<typeof Pagination>;

export const Default: Story = {
  args: {
    pageCount: 10,
    currentPage: 1
  }
};
