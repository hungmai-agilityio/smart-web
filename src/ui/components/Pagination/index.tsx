'use client';

import { useRouter, useSearchParams } from 'next/navigation';
import { Photo } from '@/ui/components/common';

interface PaginationProps {
  pageCount: number;
  currentPage: number;
}

export const Pagination = ({ pageCount, currentPage }: PaginationProps) => {
  const router = useRouter();
  const searchParams = useSearchParams();

  // Handles page index changes
  const handlePageChange = (page: number) => {
    if (page >= 1 && page <= pageCount) {
      const params = new URLSearchParams(searchParams.toString());
      params.set('portfolio-page', page.toString());
      router.push(`?${params.toString()}`, { scroll: false });
    }
  };

  // Render page item
  const renderPageNumbers = () => {
    const pages = [];

    const renderPageButton = (page: number) => (
      <button
        key={page}
        className={`px-4 py-2 ${currentPage === page ? 'bg-white text-second rounded-full drop-shadow-lg' : ''}`}
        onClick={handlePageChange.bind(null, page)}
      >
        {page}
      </button>
    );

    // Default page 1
    pages.push(renderPageButton(1));

    if (currentPage > 3) {
      pages.push(<span key="prev-page">...</span>);
    }

    const startPage = currentPage - 1 > 1 ? currentPage - 1 : 2;
    const endPage =
      currentPage + 1 < pageCount ? currentPage + 1 : pageCount - 1;

    for (let i = startPage; i <= endPage; i++) {
      pages.push(renderPageButton(i));
    }

    if (currentPage < pageCount - 2) {
      pages.push(<span key="next-page">...</span>);
    }

    if (pageCount > 1) {
      pages.push(renderPageButton(pageCount));
    }

    return pages;
  };

  return (
    <div className="flex justify-center gap-5 mt-4">
      <button
        className="px-3"
        onClick={handlePageChange.bind(null, currentPage - 1)}
        disabled={currentPage === 1}
      >
        <Photo
          src="/prev-icon.svg"
          alt="prev icon"
          fixedSize={{ width: 16, height: 16 }}
        />
      </button>
      {renderPageNumbers()}
      <button
        className="px-3"
        onClick={handlePageChange.bind(null, currentPage + 1)}
        disabled={currentPage === pageCount}
      >
        <Photo
          src="/next-icon.svg"
          alt="next icon"
          fixedSize={{ width: 16, height: 16 }}
        />
      </button>
    </div>
  );
};
