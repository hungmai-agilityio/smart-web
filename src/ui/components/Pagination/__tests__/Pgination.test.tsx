import { render, screen, fireEvent } from '@testing-library/react';
import { useRouter, useSearchParams } from 'next/navigation';
import { Pagination } from '@/ui/components/Pagination';

// Mock useRouter and useSearchParams
jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
  useSearchParams: jest.fn()
}));

describe('Pagination Component', () => {
  const push = jest.fn();
  const searchParams = new URLSearchParams();
  const pageCount = 5;
  const currentPage = 1;

  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue({ push });
    (useSearchParams as jest.Mock).mockReturnValue(searchParams);
  });

  it('should render pagination buttons correctly', () => {
    render(<Pagination pageCount={pageCount} currentPage={currentPage} />);

    expect(screen.getByAltText('prev icon')).toBeInTheDocument();
    expect(screen.getByAltText('next icon')).toBeInTheDocument();

    expect(screen.getByText('1')).toBeInTheDocument();
    expect(screen.getByText('2')).toBeInTheDocument();
    expect(screen.getByText('...')).toBeInTheDocument();
    expect(screen.getByText('5')).toBeInTheDocument();
  });

  it('disables prev button on the first page', () => {
    render(<Pagination pageCount={pageCount} currentPage={currentPage} />);

    const prevButton = screen.getByAltText('prev icon').closest('button');
    expect(prevButton).toBeDisabled();
  });

  it('should call router.push with correct URL when page changes', () => {
    render(<Pagination pageCount={pageCount} currentPage={currentPage} />);

    // Click next page button
    const nextButton = screen.getByAltText('next icon').closest('button');

    fireEvent.click(nextButton!);

    expect(push).toHaveBeenCalledWith('?portfolio-page=2', { scroll: false });
  });

  it('should be render match to snapshot', () => {
    const { container } = render(
      <Pagination pageCount={pageCount} currentPage={currentPage} />
    );

    expect(container).toMatchSnapshot();
  });
});
