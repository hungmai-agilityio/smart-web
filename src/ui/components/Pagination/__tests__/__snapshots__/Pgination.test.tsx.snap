// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Component should be render match to snapshot 1`] = `
<div>
  <div
    class="flex justify-center gap-5 mt-4"
  >
    <button
      class="px-3"
      disabled=""
    >
      <div
        class="relative w-[16px] h-[16px]"
      >
        <img
          alt="prev icon"
          class="transition-opacity duration-300"
          data-nimg="1"
          decoding="async"
          height="16"
          loading="lazy"
          src="/prev-icon.svg"
          style="color: transparent; object-fit: cover;"
          width="16"
        />
      </div>
    </button>
    <button
      class="px-4 py-2 bg-white text-second rounded-full drop-shadow-lg"
    >
      1
    </button>
    <button
      class="px-4 py-2 "
    >
      2
    </button>
    <span>
      ...
    </span>
    <button
      class="px-4 py-2 "
    >
      5
    </button>
    <button
      class="px-3"
    >
      <div
        class="relative w-[16px] h-[16px]"
      >
        <img
          alt="next icon"
          class="transition-opacity duration-300"
          data-nimg="1"
          decoding="async"
          height="16"
          loading="lazy"
          src="/next-icon.svg"
          style="color: transparent; object-fit: cover;"
          width="16"
        />
      </div>
    </button>
  </div>
</div>
`;
